ion-hash/src/lib.rs

```diff
@@ -0,0 +1,141 @@
+//!   let digest = ion_hash::sha256(&elem);
+//!   println!("{:?}", digest);
+//! # Ok(())
+//! # }
+//! ```
+
+use ion_rs::result::{illegal_operation, IonResult};
+use ion_rs::{value::Element, IonType};
+
+use sha2::{digest::Output, Digest, Sha256};
+
+// A `try`-like macro to workaround the Element API ergonomics. This API
+// requires checking the type and then calling the appropriate getter function
+// (which returns a None if you got it wrong). This macro turns the `None` into
+// an `IonError`.
+// TODO: Remove this once the Element API is reworked.
+macro_rules! t {
+    ($getter:expr) => {
+        match $getter {
+            Some(value) => value,
+            None => illegal_operation("Missing expected value")?,
+        }
+    };
+}
+
+/// Utility to hash an [`Element`] using SHA-256 as the hash function.
+pub fn sha256<E: Element>(elem: &E) -> IonResult<Vec<u8>> {
+    let mut hasher = IonHasher::new(Sha256::new());
+    let result = hasher.hash_element(elem)?;
+    Ok(Vec::from(result.as_slice()))
+}
+
+/// Bytes markers as per the spec.
+struct Markers;
+impl Markers {
+    /// single byte begin marker
+    const B: u8 = 0x0B;
+    /// single byte end marker
+    const E: u8 = 0x0E;
+    /// single byte escape
+    const ESC: u8 = 0x0C;
+    /// type qualifier octet consisting of a four-bit type code T followed by a four-bit qualifier Q
+    /// (this varies per type)
+    const TQ_STRING: u8 = 0x80;
+}
+
+// TODO: In the fullness of time, we will have a IonHashReader and a
+// IonHashWriter. This will allow reading/writing a value *while* computing a
+// hash. For now, we provide a standalone hasher using the Element API.
+/// Provides Ion hash over arbitrary [`Element`] instances with a given
+/// [`Digest`] algorithm.
+pub struct IonHasher<D>
+where
+    D: Digest,
+{
+    hasher: D,
+}
+
+impl<D> IonHasher<D>
+where
+    D: Digest,
+{
+    fn new(hasher: D) -> Self {
+        Self { hasher }
+    }
+
+    /// Computes the Ion hash over an [`Element`] recursively using this
+    /// hasher's [`Digest`]
+    fn hash_element<E: Element + ?Sized>(&mut self, elem: &E) -> IonResult<Output<D>> {
+        let serialized_bytes = match elem.ion_type() {
+            IonType::Null => todo!(),
+            IonType::Boolean => todo!(),
+            IonType::Integer => todo!(),
+            IonType::Float => todo!(),
+            IonType::Decimal => todo!(),
+            IonType::Timestamp => todo!(),
+            IonType::Symbol => todo!(),
+            IonType::String => self.visit_string(t!(elem.as_str())),
+            IonType::Clob => todo!(),
+            IonType::Blob => todo!(),
+            IonType::List => todo!(),
+            IonType::SExpression => todo!(),
+            IonType::Struct => todo!(),
+        };
+
+        // TODO: Annotations
+
+        self.hasher.update(serialized_bytes);
+        Ok(self.hasher.finalize_reset())
+    }
+
+    fn visit_string(&mut self, value: &str) -> Vec<u8> {
```

Comment:
It seems like the `Self` could have a reusable `Vec<u8>` that's updated by each visitor and reset between calls to `hash_element`.

Reply:
Agreed! I plan on reworking this as we get into other types.